src/Steps.cpp

```diff
@@ -747,6 +750,158 @@ void Steps::SetCachedMeasureInfo(const MeasureInfo ms[NUM_PLAYERS])
 	m_bAreCachedMeasureInfoJustLoaded = true;
 }
 
+const RString Steps::GetGrooveStatsHash() const
+{
+	return GrooveStatsHash;
+}
+
+int Steps::GetGrooveStatsHashVersion() const
+{
+	return GrooveStatsHashVersion;
+}
+
+void Steps::CalculateGrooveStatsHash(bool forceRecalculate)
+{
+	if (!forceRecalculate
+		&& GrooveStatsHashVersion == CURRENT_GROOVE_STATS_HASH_VERSION
+		&& m_bIsCachedGrooveStatsHashJustLoaded == true)
+	{
+		m_bIsCachedGrooveStatsHashJustLoaded = false;
+		return;
+	}
+	this->Decompress();
+
+	RString smNoteData = this->MinimizedChartString();
+
+	TimingData * timingData = this->GetTimingData();
+	std::vector<TimingSegment *> segments = timingData->GetTimingSegments(SEGMENT_BPM);
+	std::vector<RString> bpmStrings;
+	for (TimingSegment *segment : segments)
+	{
+		BPMSegment *bpmSegment = ToBPM(segment);
+		float beat = bpmSegment->GetBeat();
+		float bpm = bpmSegment->GetBPM();
+		std::ostringstream os;
+		os << std::fixed << std::setprecision(3) << beat;
```

Comment:
As per the other comment thread, I believe this is where the rounding error is occurring. `setprecision` is truncating based on a value ITGm has already processed. Specifically ITGm can turn a beat from `236.813` in the ssc file to `236.8125`, which truncates instead of rounding to get back to the original value.

Reply:
I think the way to do this for would be the following. This is similar to what we're doing in the lua [here](https://github.com/Simply-Love/Simply-Love-SM5/blob/itgmania-release/Scripts/SL-ChartParser.lua#L117-L126):

```c++
std::string NormalizeDecimal(float num) {
  float mult = 1000.0;
  float rounded = (num * mult + 0.5 - std::fmod(num * mult + 0.5, 1)) / mult;
  std::ostringstream os;
  os << std::fixed << std::setprecision(3) << rounded;
  return os.str();
}
```

My test case for this behavior is `82.0625` where we expect it to return `82.063` while the current implementation returns `82.062`.

Reply:
Fwiw IIUC using `std::fixed` along with `setprecision(3)` is functionally the same as "%.3f" which does do one of the many [types of rounding](https://en.wikipedia.org/wiki/Rounding#Rounding_to_the_nearest_integer), but the not the variant of rounding we're looking for here. This implementation would have been fine for V2 version of the hashing algorithm but not the current (V3) version.